content/articles/fcm-with-spring-and-angular/index.md

```diff
@@ -0,0 +1,500 @@
+
+#### What will happen on the client-side?
+
+Client-side, we will either show the message within the app or as a notification popup. The former when our application is opened and the latter when our application is closed. Although you may be asking: **how can we show the notification when our application is inactive?** We would do this with the help of a **service worker**.
+
+##### What is a service worker?
+
+A service worker is a special script that runs on a separate thread from your application. It allows you to intercept requests, cache data for offline use, and in our case, send notifications. Since a service worker is separate from our app, we can use it to send notifications while the app is inactive. Although, when our app is active, we will let our angular project handle the message by displaying it on the page.
+
+#### What will happen on the server-side?
+
+Server-side, our Spring Boot app will use an SDK from Firebase called **the admin SDK**. This SDK allows our application to interact with Firebase so we can ask it to send notifications for us. When we initialize Firebase in the Firebase console, they will give us a special JSON file. We will use this JSON file to authorize our Spring Boot app to send notifications.
+
+Whenever we want to send notifications, we have to create a `Message` object. This will contain all the info about the notification we want to send. This will include the title, description, an icon URL, and any platform-specific information. We have two ways to send the notification, as a topic or direct notification. 
+    
+##### Topic notifications vs direct notifications
+
+A topic notification is a notification with a specified tag known as a **topic**. Users will subscribe to be notified of any messages with a topic of their choice. Whenever a user subscribes to a topic, they will send our Spring Boot app their token and the name of the topic to subscribe to. Using this, we can tell Firebase to send them notifications about that topic.
+
+Additionally, we also have the option to send a direct notification. Here, we specify in the `Message` object the token of the user to notify. Then, Firebase will send the notification to that particular user.
+
+#### Illustrating the full architecture
+
+To illustrate how this entire project will work, here is a handy flow chart I made to picture it. Hopefully, this will clear up any confusion you may have had about the architecture:
+
+![a flow chart to illustrate the architecture](flow-chart.png)
```

Comment:
`flow-chart.png`

You must specify full path for the image. For example, it must be `/engineering-education/[ARTICLE-DIRECTORY-NAME]/[IMAGE-NAME]`

Do this change for all the images that you have used.